### PR TITLE
Add retry logic to Azure REST calls.

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
Sometimes the REST APIs fail due to transient issues.  I've added some
retry logic, the same that we use in Helix.

### CODE REVIEW

**Comment by jhendrixMSFT** [comment]
/cc @ChadNedzlek @karajas @chcosta 
This is a port of saferequests.py from Helix.

File: src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs [hunk 1]
```diff
@@ -182,6 +186,95 @@ public static List<string> GetHeaderValues(HttpRequestHeaders headers, string he
...
+            // add a bit of randomness to the retry delay
+            var rng = new Random();
+
+            while (retries < retryCount)
+            {
+                if (retries > 1)
+                {
+                    int delay = retryDelaySeconds * (retries - 1) * rng.Next(1, 5);
+                    loggingHelper.LogMessage("Waiting {0} seconds before retry", delay);
+                    Thread.Sleep(delay * 1000);
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs]: await Task.Delay()

File: src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs [hunk 2]
```diff
@@ -182,6 +186,95 @@ public static List<string> GetHeaderValues(HttpRequestHeaders headers, string he
...
+
+            // add a bit of randomness to the retry delay
+            var rng = new Random();
+
+            while (retries < retryCount)
+            {
+                if (retries > 1)
+                {
+                    int delay = retryDelaySeconds * (retries - 1) * rng.Next(1, 5);
+                    loggingHelper.LogMessage("Waiting {0} seconds before retry", delay);
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs]: You'll probably want to do a lot of this logging at "low" level

File: src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs [hunk 3]
```diff
@@ -182,6 +186,95 @@ public static List<string> GetHeaderValues(HttpRequestHeaders headers, string he
...
+                        return response;
+                    }
+                }
+
+                ++retries;
+            }
+
+            // retry count exceeded
+            loggingHelper.LogWarning("Retry count {0} exceeded", retryCount);
+            throw new Exception(string.Format("Request failed with status {0} response {1}",
```
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs]: I wish we could throw something more useful.  Forcing people to do "catch(Exception)" isn't ideal.

**Comment by ChadNedzlek** [comment]
Just a few minor comments, otherwise fine.

File: src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs [hunk 4]
```diff
@@ -182,6 +186,95 @@ public static List<string> GetHeaderValues(HttpRequestHeaders headers, string he
...
         }
 
+        private static bool IsWithinRetryRange(HttpStatusCode statusCode)
+        {
+            // Retry on http 'server error' codes (5xx) as well as redirect
+
+            var rawStatus = (int)statusCode;
+            if (rawStatus == 302)
+                return true;
+            else if (rawStatus >= 500 && rawStatus <= 599)
```
Comment (karajas) [inline src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs]: I believe we've also seen errors on the client side with status codes from 400-499. Do we want to retry on these as well?

**Comment by schaabs** [comment]
Seems like all these calls are not resilient to retry.  Syncing these changes and rebuilding I get the following error on retry:

    d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error : The request message was already sent. Cannot send the same request message multiple times.\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error :    at System.Net.Http.HttpClient.CheckRequestMessage(HttpRequestMessage request)\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error :    at System.Net.Http.HttpClient.SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error :    at System.Net.Http.HttpClient.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error :    at Microsoft.DotNet.Build.CloudTestTasks.UploadClient.<>c__DisplayClass3_4.<UploadBlockBlobAsync>b__1() in D:\git\buildtools2\src\Microsoft.DotNet.Build.CloudTestTasks\UploadClient.cs:line 169\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error :    at Microsoft.DotNet.Build.CloudTestTasks.AzureHelper.<RequestWithRetry>d__11.MoveNext() in D:\git\buildtools2\src\Microsoft.DotNet.Build.CloudTestTasks\AzureHelper.cs:line 239\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error : --- End of stack trace from previous location where exception was thrown ---\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error :    at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error :    at Microsoft.DotNet.Build.CloudTestTasks.UploadClient.<UploadBlockBlobAsync>d__3.MoveNext() in D:\git\buildtools2\src\Microsoft.DotNet.Build.CloudTestTasks\UploadClient.cs:line 169\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error : --- End of stack trace from previous location where exception was thrown ---\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.GetResult()\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error :    at Microsoft.DotNet.Build.CloudTestTasks.UploadToAzure.<UploadAsync>d__25.MoveNext() in D:\git\buildtools2\src\Microsoft.DotNet.Build.CloudTestTasks\UploadToAzure.cs:line 157\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error : --- End of stack trace from previous location where exception was thrown ---\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error :    at System.Runtime.CompilerServices.TaskAwaiter.GetResult()\r [d:\dotnet-reliability\test\genstress.proj]
       d:\dotnet-reliability\Tools\CloudTest.targets(288,5): error :    at Microsoft.DotNet.Build.CloudTestTasks.UploadToAzure.<ExecuteAsync>d__24.MoveNext() in D:\git\buildtools2\src\Microsoft.DotNet.Build.CloudTestTasks\UploadToAzure.cs:line 126 [d:\dotnet-reliability\test\genstress.proj]

File: src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs (hunk 1 above)
Comment (jhendrixMSFT) [inline src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs]: Will fix.

File: src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs (hunk 4 above)
Comment (jhendrixMSFT) [inline src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs]: I don't see anything in the Helix code that retries on these failures (RETRY_RANGE_CLIENT_ERROR).  I can add it however I'd like to see some precedent before blindly adding.  @ChadNedzlek what's your take.

**Comment by jhendrixMSFT** [comment]
@schaabs Thanks for testing this out and sharing the failure, I'm looking into it now.

File: src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs (hunk 4 above)
Comment (ChadNedzlek) [inline src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs]: 4xx errors are _usually_ less transient. But I do feel like we've seen on at least one occasion a 404 that wasn't accurate.  4xx error _definitely_ aren't success, so it shouldn't... hurt anything.

Comment (jhendrixMSFT) [inline src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs]: OK I'll add them to the range.

File: src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs (hunk 3 above)
Comment (jhendrixMSFT) [inline src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs]: Good point, I'll change it to HttpRequestException.

**Comment by jhendrixMSFT** [comment]
@schaabs This should be fixed now, could you give it another go?

**Pull request merged**